File: 作业2-蒙特卡洛抽样/monte_carlo.py
"""
monte_carlo.py

实现说明：
- 读取 `versatile_params.csv`（每行：序号,a,b,c），共25组
- 根据预测值到分箱的映射（25等宽分箱，宽度=0.04），查表获取 (a,b,c)
- 使用解析反函数对每个时刻做 N 次蒙特卡洛抽样：
    x = CDF^{-1}(u|a,b,c) = c - (1/a) * ln(u^{-1/b} - 1)
  其中 u ~ Uniform(0,1)
- 将 N 条场景保存为 CSV，计算每条场景的 2 小时区间最大-最小值（DeltaP）并保存统计信息
- 计算解析 90% 置信区间（u=0.05 和 u=0.95 的反函数）并验证抽样点落在区间内的比例

输出文件：
- monte_carlo_scenarios_N{N}.csv
- monte_carlo_summary_N{N}.txt
- deltaP_hist_N{N}.png (如果 matplotlib 可用)
"""

from pathlib import Path
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

ROOT = Path(__file__).parent
PARAM_CSV = ROOT / "versatile_params.csv"
PW_FORECAST_CSV = ROOT / "Pw_2024-10-01.csv"

# Past values for 2024-10-01 06:00-07:45
PAST_P = [0.85, 0.823, 0.711, 0.774, 0.854, 0.918, 0.914, 0.881]
# Forecast values for 2024-10-01 08:00-09:45 (8 points)
FORECAST_P = [0.885, 0.868, 0.776, 0.668, 0.611, 0.594, 0.558, 0.452]
TRUE_P = [0.901, 0.834, 0.709, 0.632, 0.609, 0.564, 0.487, 0.264]
TIME_LABELS = [
    "2024-10-01 08:00",
    "2024-10-01 08:15",
    "2024-10-01 08:30",
    "2024-10-01 08:45",
    "2024-10-01 09:00",
    "2024-10-01 09:15",
    "2024-10-01 09:30",
    "2024-10-01 09:45",
]
BIN_COUNT = 25
BIN_WIDTH = 1.0 / BIN_COUNT

def load_params(csv_path: Path):
    df = pd.read_csv(csv_path)
    # expect columns: 序号,a,b,c or index starting at 1
    # return dict: index (1-based) -> (a,b,c)
    mapping = {}
    for _, row in df.iterrows():
        idx = int(row[df.columns[0]])
        a = float(row[df.columns[1]])
        b = float(row[df.columns[2]])
        c = float(row[df.columns[3]])
        mapping[idx] = (a, b, c)
    return mapping

def value_to_bin_index(v: float) -> int:
    # bins: [0,0.04) -> 1, [0.04,0.08) ->2, ..., [0.96,1.0] ->25
    if v >= 1.0:
        return BIN_COUNT
    if v < 0.0:
        return 1
    idx = int(v / BIN_WIDTH) + 1
    idx = max(1, min(BIN_COUNT, idx))
    return idx

def inv_cdf(u: np.ndarray, a: float, b: float, c: float) -> np.ndarray:
    # Handle edge u values: clamp to (eps, 1-eps)
    eps = 1e-12
    u = np.clip(u, eps, 1 - eps)
    # formula: c - (1/a) * ln(u^{-1/b} - 1)
    # compute u^{-1/b} safely
    with np.errstate(divide='ignore', invalid='ignore'):
        power = np.power(u, -1.0 / b)
        inner = power - 1.0
        # ensure inner>0 to avoid log domain errors; 
        # inner should be >0 because power>1 for u in (0,1)
        inner = np.clip(inner, 1e-300, None)
        vals = c - (1.0 / a) * np.log(inner)
    return vals

def sample_scenarios(params_map, N=1000, seed=None):
    rng = np.random.default_rng(seed)
    M = len(FORECAST_P)
    scenarios = np.zeros((N, M), dtype=float)
    ci_bounds = []
    for j, p in enumerate(FORECAST_P):
        idx = value_to_bin_index(p)
        if idx not in params_map:
            raise KeyError(f"Bin index {idx} not found in params table")
        a, b, c = params_map[idx]
        u = rng.random(N)
        samples = inv_cdf(u, a, b, c).round(4)
        scenarios[:, j] = samples
        lo = inv_cdf(0.05, a, b, c).round(4)
        hi = inv_cdf(0.95, a, b, c).round(4)
        ci_bounds.append((float(lo), float(hi)))
    return scenarios, ci_bounds

def analyze_and_save(scenarios, ci_bounds, N, out_folder):
    # scenarios: (N, M)
    df = pd.DataFrame(scenarios, columns=TIME_LABELS)
    scen_csv = out_folder / f"scenarios_N{N}.csv"
    df.to_csv(scen_csv, index=False)

    # proportion of sampled points within analytical 90% CI
    total_points = scenarios.size
    inside = 0
    per_time_props = []
    for j in range(scenarios.shape[1]):
        lo, hi = ci_bounds[j]
        col = scenarios[:, j]
        inside_j = np.sum((col >= lo) & (col <= hi))
        per_time_props.append(float(inside_j / scenarios.shape[0]))
        inside += inside_j
    overall_prop = float(inside / total_points)

    return {
        'scenarios_csv': str(scen_csv),
        'per_time_props': per_time_props,
        'overall_prop': overall_prop,
    }

def convergence_test(params_map, Ns=(100, 1000, 5000, 10000), seed=42):
    results = []
    for N in Ns:
        scenarios, _ = sample_scenarios(params_map, N=N, seed=seed)
        mean_profile = scenarios.mean(axis=0)
        diff = mean_profile - np.array(FORECAST_P)
        results.append({'N': N, 'mean_profile': mean_profile, 'diff': diff})
    return results

def plot_scenarios(scenarios, ci_bounds, N, out_folder, true_series=None):
    # full timeline: 00:00 - 09:45 by 15 minutes
    full_idx = pd.date_range(start='2024-10-01 06:00:00', end='2024-10-01 09:45:00', freq='15min')
    scenario_len = scenarios.shape[1]
    scenario_start_pos = len(full_idx) - scenario_len
    x = np.arange(len(full_idx))

    plt.figure(figsize=(12, 6))

    # Part 1: plot all true values for 06:00-09:45 (if provided)
    if true_series is not None:
        true_vals = true_series.reindex(full_idx)
        plt.plot(x, true_vals, color='red', marker='o', linewidth=1.5, label='True (06:00-09:45)')

    # Part 2: plot all scenarios (only on 08:00-09:45 positions)
    for i in range(scenarios.shape[0]):
        y = np.full(len(full_idx), np.nan)
        y[scenario_start_pos:scenario_start_pos + scenario_len] = scenarios[i]
        plt.plot(x, y, color='C0', alpha=0.1, linewidth=0.6)

    # Part 3: 90% CI (08:00-09:45)
    ci_lo = np.full(len(full_idx), np.nan)
    ci_hi = np.full(len(full_idx), np.nan)
    ci_lo[scenario_start_pos:scenario_start_pos + scenario_len] = [b[0] for b in ci_bounds]
    ci_hi[scenario_start_pos:scenario_start_pos + scenario_len] = [b[1] for b in ci_bounds]
    plt.fill_between(x, ci_lo, ci_hi, color='orange', alpha=0.5, label='90% CI (08:00-09:45)')

    # Part 4: Forecast values (08:00-09:45)
    fore = np.full(len(full_idx), np.nan)
    fore[scenario_start_pos:scenario_start_pos + scenario_len] = FORECAST_P
    plt.plot(x, fore, color='blue', marker='s', linestyle='--', linewidth=1.5, label='Forecast (08:00-09:45)')

    # x ticks: show each hour
    xticks_pos = []
    xticks_labels = []
    for i, t in enumerate(full_idx):
        xticks_pos.append(i)
        xticks_labels.append(t.strftime('%H:%M'))
    plt.xticks(xticks_pos, xticks_labels, rotation=0)
    plt.xlim(0, len(full_idx) - 1)
    plt.xlabel('Time (06:00 - 09:45)')
    plt.ylabel('Value')
    plt.title(f'Monte Carlo Scenarios (N={N}) — True, Scenarios, 90% CI, Forecast')
    plt.legend()
    plt.tight_layout()
    plot_path = out_folder / f"scenarios_plot_N{N}.png"
    plt.savefig(plot_path)
    plt.close()
    return str(plot_path)

def main():
    N, seed = 1000, 12345
    params_map = load_params(PARAM_CSV)
    scenarios, ci_bounds = sample_scenarios(params_map, N, seed)
    output_folder = Path('output')
    out = analyze_and_save(scenarios, ci_bounds, N, output_folder)
    # Build true_series from PAST_P (06:00-07:45) and TRUE_P (08:00-09:45)
    full_idx = pd.date_range(start='2024-10-01 06:00:00', end='2024-10-01 09:45:00', freq='15min')
    true_vals = np.full(len(full_idx), np.nan)
    # PAST_P covers 06:00-07:45 -> first 8 positions
    true_vals[0:len(PAST_P)] = PAST_P
    # TRUE_P covers 08:00-09:45 -> last 8 positions
    start_pos = len(full_idx) - len(TRUE_P)
    true_vals[start_pos:start_pos + len(TRUE_P)] = TRUE_P
    true_series = pd.Series(index=full_idx, data=true_vals)
    plot_path = plot_scenarios(scenarios, ci_bounds, N, output_folder, true_series=true_series)

    print('Outputs:')
    print(' - scenarios CSV:', out['scenarios_csv'])
    print(' - scenarios plot PNG:', plot_path)
    print(' - Proportion of points within 90% CI per time point:', out['per_time_props'])
    print(' - Overall proportion of points within 90% CI:', out['overall_prop'])

    # Convergence quick check (small set)
    conv = convergence_test(params_map, Ns=(100, 1000), seed=seed)
    print('\nConvergence check (mean - forecast) for Ns=100,1000:')
    for r in conv:
        diffs = r['diff']
        with open (output_folder / f"summary_N{r['N']}.txt", 'w') as f:
            f.write(f"N={r['N']}:\n")
            f.write(f"Mean profile: {r['mean_profile']}\n")
            f.write(f"Difference from forecast: {diffs}\n")

if __name__ == '__main__':
    main()

File: 作业2-蒙特卡洛抽样/monte_carlo.m
% monte_carlo.m
function monte_carlo_main()
    clear; clc;
    
    % --- 1. 参数与常量定义 ---
    % 定义时间节点字符串，用于生成CSV表头
    TIME_LABELS = {
        'T2024_10_01_08_00', 'T2024_10_01_08_15', 'T2024_10_01_08_30', ...
        'T2024_10_01_08_45', 'T2024_10_01_09_00', 'T2024_10_01_09_15', ...
        'T2024_10_01_09_30', 'T2024_10_01_09_45'};

    PAST_P = [0.85, 0.823, 0.711, 0.774, 0.854, 0.918, 0.914, 0.881];
    FORECAST_P = [0.885, 0.868, 0.776, 0.668, 0.611, 0.594, 0.558, 0.452];
    TRUE_P = [0.901, 0.834, 0.709, 0.632, 0.609, 0.564, 0.487, 0.264];
    
    BIN_COUNT = 25;
    BIN_WIDTH = 1.0 / BIN_COUNT;
    
    N = 1000;
    seed_val = 12345;
    
    % 创建输出文件夹
    out_folder = 'output_2';
    if ~exist(out_folder, 'dir')
        mkdir(out_folder);
    end
    
    % --- 2. 核心逻辑执行 ---
    % 读取分布参数
    params_map = load_params('versatile_params.csv');
    
    % 蒙特卡洛抽样
    [scenarios, ci_bounds] = sample_scenarios(params_map, FORECAST_P, N, seed_val, BIN_COUNT, BIN_WIDTH);
    
    % 统计与保存CSV
    out = analyze_and_save(scenarios, ci_bounds, N, TIME_LABELS, out_folder);
    
    % 绘图
    plot_path = plot_scenarios(scenarios, ci_bounds, N, PAST_P, TRUE_P, FORECAST_P, out_folder);
    
    % 输出终端日志
    fprintf('Outputs:\n');
    fprintf(' - scenarios CSV: %s\n', out.scenarios_csv);
    fprintf(' - scenarios plot PNG: %s\n', plot_path);
    fprintf(' - Proportion of points within 90%% CI per time point:\n');
    disp(out.per_time_props);
    fprintf(' - Overall proportion of points within 90%% CI: %.4f\n\n', out.overall_prop);
    
    % 收敛性测试
    Ns_test = [100, 1000];
    conv_results = convergence_test(params_map, FORECAST_P, Ns_test, seed_val, BIN_COUNT, BIN_WIDTH);
    fprintf('Convergence check (mean - forecast) for Ns=100,1000:\n');
    for i = 1:length(conv_results)
        r = conv_results(i);
        txt_path = fullfile(out_folder, sprintf('summary_N%d.txt', r.N));
        fid = fopen(txt_path, 'w');
        fprintf(fid, 'N=%d:\n', r.N);
        fprintf(fid, 'Mean profile: %s\n', mat2str(r.mean_profile, 4));
        fprintf(fid, 'Difference from forecast: %s\n', mat2str(r.diff, 4));
        fclose(fid);
        
        fprintf('N=%d:\n', r.N);
        fprintf('Mean profile: %s\n', mat2str(r.mean_profile, 4));
        fprintf('Difference: %s\n\n', mat2str(r.diff, 4));
    end
end

% =========================================================================
% Local Functions
% =========================================================================

function params = load_params(csv_path)
    % 读取参数矩阵，假设格式：序号, a, b, c
    M = readmatrix(csv_path);
    params = zeros(25, 3);
    for i = 1:size(M, 1)
        idx = M(i, 1);
        params(idx, 1) = M(i, 2); % a
        params(idx, 2) = M(i, 3); % b
        params(idx, 3) = M(i, 4); % c
    end
end

function idx = value_to_bin_index(v, BIN_COUNT, BIN_WIDTH)
    if v >= 1.0
        idx = BIN_COUNT;
        return;
    end
    if v < 0.0
        idx = 1;
        return;
    end
    idx = floor(v / BIN_WIDTH) + 1;
    idx = max(1, min(BIN_COUNT, idx));
end

function vals = inv_cdf(u, a, b, c)
    eps_val = 1e-12;
    % 将u限制在合理范围内
    u = max(eps_val, min(1 - eps_val, u));
    
    power_val = u .^ (-1.0 / b);
    inner = power_val - 1.0;
    % 确保对数内的值大于0
    inner = max(1e-300, inner);
    vals = c - (1.0 / a) * log(inner);
end

function [scenarios, ci_bounds] = sample_scenarios(params_map, FORECAST_P, N, seed_val, BIN_COUNT, BIN_WIDTH)
    rng(seed_val, 'twister'); % 设定随机数种子
    M = length(FORECAST_P);
    scenarios = zeros(N, M);
    ci_bounds = zeros(M, 2);
    
    for j = 1:M
        p = FORECAST_P(j);
        idx = value_to_bin_index(p, BIN_COUNT, BIN_WIDTH);
        
        a = params_map(idx, 1);
        b = params_map(idx, 2);
        c = params_map(idx, 3);
        
        % 抽样 N 个 u ~ Uniform(0,1)
        u = rand(N, 1);
        samples = round(inv_cdf(u, a, b, c), 4);
        scenarios(:, j) = samples;
        
        % 90% 置信区间
        lo = round(inv_cdf(0.05, a, b, c), 4);
        hi = round(inv_cdf(0.95, a, b, c), 4);
        ci_bounds(j, :) = [lo, hi];
    end
end

function out = analyze_and_save(scenarios, ci_bounds, N, TIME_LABELS, out_folder)
    % 1. 保存为 CSV
    scen_csv = fullfile(out_folder, sprintf('scenarios_N%d.csv', N));
    T = array2table(scenarios, 'VariableNames', TIME_LABELS);
    writetable(T, scen_csv);
    
    % 2. 验证抽样点落在 90% 解析置信区间内的比例
    total_points = numel(scenarios);
    inside_total = 0;
    M = size(scenarios, 2);
    per_time_props = zeros(1, M);
    
    for j = 1:M
        lo = ci_bounds(j, 1);
        hi = ci_bounds(j, 2);
        col = scenarios(:, j);
        
        inside_j = sum(col >= lo & col <= hi);
        per_time_props(j) = inside_j / N;
        inside_total = inside_total + inside_j;
    end
    overall_prop = inside_total / total_points;
    
    out.scenarios_csv = scen_csv;
    out.per_time_props = per_time_props;
    out.overall_prop = overall_prop;
end

function results = convergence_test(params_map, FORECAST_P, Ns, seed_val, BIN_COUNT, BIN_WIDTH)
    results = struct('N', {}, 'mean_profile', {}, 'diff', {});
    for i = 1:length(Ns)
        N = Ns(i);
        [scenarios, ~] = sample_scenarios(params_map, FORECAST_P, N, seed_val, BIN_COUNT, BIN_WIDTH);
        mean_profile = mean(scenarios, 1);
        diff_val = mean_profile - FORECAST_P;
        
        results(i).N = N;
        results(i).mean_profile = mean_profile;
        results(i).diff = diff_val;
    end
end

function plot_path = plot_scenarios(scenarios, ci_bounds, N, PAST_P, TRUE_P, FORECAST_P, out_folder)
    % 构建完整的时间序列: 06:00 到 09:45，步长15分钟
    t_start = datetime(2024, 10, 1, 6, 0, 0);
    t_end = datetime(2024, 10, 1, 9, 45, 0);
    full_idx = t_start : minutes(15) : t_end;
    
    fig = figure('Position', [100, 100, 1000, 500], 'Visible', 'off');
    hold on;
    
    % 数据段落对齐
    len_past = length(PAST_P);
    len_future = length(FORECAST_P);
    idx_future = (length(full_idx) - len_future + 1) : length(full_idx);
    x_future = full_idx(idx_future);
    
    % 1. 绘制所有的 Scenarios (为了视觉效果，使用非常浅的颜色和细线)
    % 使用 rgba 控制透明度，R2021a+ 支持
    color_scenario = [0, 0.4470, 0.7410, 0.05]; 
    h_scen = plot(x_future, scenarios', 'Color', color_scenario, 'LineWidth', 0.5);
    
    % 2. 绘制 90% 置信区间 (08:00 - 09:45)
    ci_lo = ci_bounds(:, 1)';
    ci_hi = ci_bounds(:, 2)';
    x_fill = [x_future, fliplr(x_future)];
    y_fill = [ci_lo, fliplr(ci_hi)];
    h_ci = fill(x_fill, y_fill, [1, 0.647, 0], 'FaceAlpha', 0.5, 'EdgeColor', 'none');
    
    % 3. 绘制真实值 (06:00 - 09:45)
    true_vals = NaN(1, length(full_idx));
    true_vals(1:len_past) = PAST_P;
    true_vals(idx_future) = TRUE_P;
    h_true = plot(full_idx, true_vals, '-ro', 'LineWidth', 1.5, 'MarkerFaceColor', 'r');
    
    % 4. 绘制预测值 (08:00 - 09:45)
    h_fore = plot(x_future, FORECAST_P, '--bs', 'LineWidth', 1.5, 'MarkerFaceColor', 'b');
    
    % 格式化图表
    xticks(full_idx);
    xtickformat('HH:mm'); % 仅保留时:分
    xlim([full_idx(1), full_idx(end)]);
    xlabel('Time (06:00 - 09:45)');
    ylabel('Value');
    title(sprintf('Monte Carlo Scenarios (N=%d) — True, Scenarios, 90%% CI, Forecast', N));
    
    % 图例 (只取一根 scenario 曲线用于显示图例)
    legend([h_true, h_scen(1), h_ci, h_fore], ...
        {'True (06:00-09:45)', 'Scenarios', '90% CI (08:00-09:45)', 'Forecast (08:00-09:45)'}, ...
        'Location', 'best');
    
    grid on;
    hold off;
    
    % 保存并关闭图像
    plot_path = fullfile(out_folder, sprintf('scenarios_plot_N%d.png', N));
    saveas(fig, plot_path);
    close(fig);
end